### PR TITLE
Removed redundancy in frontend configs by creating a core config file, added new settings.

### PR DESCRIPTION
**Pull request by archshift**
I also took the liberty to make the default INIs inlined into the `config.cpp` file instead of as a separate header, and I changed the comments on the INIs to use a semicolon in order to avoid the parse errors.

### CODE REVIEW

**Comment by yuriks** [comment]
How do we handle saving with IniH? The Qt UI will eventually have dialogs for changing the options, and we'll want to save that back to file. Does it support saving? (And does it keep comments and file layout, in that case?) I'm also hoping that there's a way of getting rid of the redundant copy of the default values in the form of the file template, since they're already given on the reading functions.

**Comment by archshift** [comment]
Rebased. 

@yuriks, INIH can definitely save the config, but I'll have to double-check about preserving the file layout, etc.

**Comment by archshift** [comment]
I've added some new settings to the PR, mainly config settings for:
- Autoplaying a game (Qt frontend), 
- Auto-loading a game file (core),
- Auto-loading a map file (Qt frontend)

**Comment by purpasmart96** [comment]
I like the option to set your game path, makes navigating for your roms quicker.

File: src/citra/config.cpp [hunk 1]
```diff
@@ -4,12 +4,33 @@
...
+pad_b =
+pad_x =
+pad_y =
+pad_r =
+pad_l =
+pad_sup =
+pad_sdown =
+pad_sleft =
+pad_sright =
+)";
```
Comment (bunnei) [inline src/citra/config.cpp]: Why are just the pad controls here?

Comment (bunnei) [inline src/citra/config.cpp]: Oh duh... because these are GLFW-specific

File: src/core/config.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+
+[Data Storage]
+# Whether to create a virtual SD card.
+# 1 (default): Yes, 0: No
+use_virtual_sd =
+
+
+[Miscellaneous]
+# Path for to a game file to be autoloaded by Citra.
```
Comment (bunnei) [inline src/core/config.cpp]: "Path to a game"...

File: src/citra_qt/settings.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2014 Citra Emulator Project
```
Comment (bunnei) [inline src/citra_qt/settings.cpp]: New files, copyright 2015?

**Comment by bunnei** [comment]
Looks fine to me. Would like a review from @neobrain though because he has a stronger opinion about changes to any of the Qt frontend code :)

**Comment by neobrain** [comment]
@archshift Please don't try to slip in unrelated changes just because they affect the same code. There's "git add -p" to separate out unrelated changes into their own commits.

**Comment by neobrain** [comment]
Concerning the main change of this PR: I don't think it's going into the right direction, for a simple reason: Setting up configuration is none of the emulation core's business. This probably needs some explanation. Note that I've reasoned about this in Gekko for a while and that we've ran into a *lot* issues with configuration handling in Dolphin, so I'm not pulling these arguments out of nowhere.

- How does the core know how to store and access information? Should it use the Windows registry? should it use a plain ini file? should it use a binary database? All of these might be valid choices which the GUI might make, and the emulation core shouldn't intervene with that.
- If configuration is represented by a file, how does the core know where to store and access information? Will the file just be placed next to the executable (aka portable builds), or will it be stored in an AppData-like directory? Again, both are valid choices from a frontend perspective - and might actually cause critical conflicts if different frontends try to access the same config file in a different way.
- How do we expect GUIs to deal with the core's setting up a basically arbitrary configuration for them beforehand? Some frontend might have a different "opinion" on what a good "default" configuration is - so to provide this default configuration effectively all settings need to be copy-pasted again, anyway, since technically there's no way for the GUI to be sure what default settings the core sets up.
- The split between the "core-config.ini" and multiple backend inis (which was already kind of artificial before this change) now are set in stone by design. There's real reason for doing this. Not only do I find this design less than ideal, but it also overcomplicates the configuration loading process, which is now split across the core and the GUI.

Mind you, this is done with the expectation that eventually Citra's core should be decoupled from the existing GUIs well enough that other people can start writing frontends for Citra in a clean way. While this is probably utopic from the current standpoint, I don't think we should make explicit misdesigns which will prevent this in the future. Especially when the only reason for it is removing well-justified redundance.


Also note that I probably should have brought up this issue before. To be frank, it has all happened in Gekko already and I didn't want to get into this discussion again initially, and I also didn't want to require another rewrite of the suggested configuration files PR. In retrospect, that was probably a mistake.

Since this comment is already fairly long, I'll make a followup comment providing some more constructive suggestions as to how to proceed.

**Comment by archshift** [comment]
>How does the core know how to store and access information...the emulation core shouldn't intervene with that.
>How does the core know where to store and access information?...might actually cause critical conflicts if different frontends try to access the same config file in a different way.

I see no way to reconcile these two ideas except replicating ***every single config option*** between the two frontends. This, IMO, is both a lazy solution and an unsustainable one. Duplicating what may become over fifty different configuration options just for the sake of *not* unifying them, to me, is a little bit silly. 

>Should it use the Windows registry? should it use a plain ini file? should it use a binary database?
>might actually cause critical conflicts if different frontends try to access the same config file in a different way.

This is already established in this PR, it uses an INI file with an already-established parser, INIH.

>How do we expect GUIs to deal with the core's setting up a basically arbitrary configuration for them beforehand? Some frontend might have a different "opinion" on what a good "default" configuration is - so to provide this default configuration effectively all settings need to be copy-pasted again, anyway, since technically there's no way for the GUI to be sure what default settings the core sets up.

I can think of an easy solution to this, which is to have the frontend pass in its own default settings to the Core config manager. If you want me to do this, let me know. I'm not sure where you're getting that they'd need to be copy-pasted again. 

>The split between the "core-config.ini" and multiple backend inis (which was already kind of artificial before this change) now are set in stone by design. There's real reason for doing this.

Could you clarify this? I'm not sure what you mean. 

>Not only do I find this design less than ideal, but it also overcomplicates the configuration loading process, which is now split across the core and the GUI.

Config loading is still controlled by the GUI, but the core config file is still wrapped up in Core. While this may not be ideal, I can't think of a better option that doesn't involve duplicating the options, which IMO is even more overcomplicated.

>Mind you, this is done with the expectation that eventually Citra's core should be decoupled from the existing GUIs well enough that other people can start writing frontends for Citra in a clean way. While this is probably utopic from the current standpoint, I don't think we should make explicit misdesigns which will prevent this in the future.

I totally agree that Citra's core should eventually be decoupled. A few notes though:
- This doesn't at all prevent Citra's core from being decoupled (the Core already uses the INIH parser, not something specified by the frontend)
- This doesn't at all prevent the startup of new, clean frontends (same reason as above)
- There's no better place to put a unified configuration loader (Common? Maybe.)

---

I hope we can come to a solution that does not involve that kind of unnecessary duplication, and I would love to hear your suggestions.

**Comment by yuriks** [comment]
I think this whole conundrum can be solved by simply making the common configuration be loaded by library code (in Common, most likely) that is invoked by both frontends. I also agree that having the config code in Core is a bit messy.

**Comment by neobrain** [comment]
>>How does the core know how to store and access information...the emulation core shouldn't intervene with that.
>>How does the core know where to store and access information?...might actually cause critical conflicts if different frontends try to access the same config file in a different way.

>I see no way to reconcile these two ideas except replicating ***every single config option*** between the two frontends. This, IMO, is both a lazy solution and an unsustainable one. Duplicating what may become over fifty different configuration options just for the sake of *not* unifying them, to me, is a little bit silly. 

My claim is exactly that there is no sane way to unify them without also making lots of trade-offs with regards to design and flexibility. I consider the replication to be worth these two points.


>>Should it use the Windows registry? should it use a plain ini file? should it use a binary database?
>>might actually cause critical conflicts if different frontends try to access the same config file in a different way.

> This is already established in this PR, it uses an INI file with an already-established parser, INIH.

That's what it's doing in the GLFW-backend, not in Qt. It uses Ini files in Qt as well, but that's not necessarily what we want to go with in the future (and since this is usually a controversial topic - please let's not derail this discussion :) whether or not we want that is besides the point anyway). Let's not rule out potential changes like that just due to a perceived redundancy.

>>How do we expect GUIs to deal with the core's setting up a basically arbitrary configuration for them beforehand? Some frontend might have a different "opinion" on what a good "default" configuration is - so to provide this default configuration effectively all settings need to be copy-pasted again, anyway, since technically there's no way for the GUI to be sure what default settings the core sets up.

>I can think of an easy solution to this, which is to have the frontend pass in its own default settings to the Core config manager. If you want me to do this, let me know. I'm not sure where you're getting that they'd need to be copy-pasted again. 

How is having the frontend pass its own default settings less redundancy than we have now?

The "copy-pasted" thing was worded poorly. What I was getting at is that with your proposed solution, all backends would then provide their own default settings mappings, which would be just as redundant as the current situation.


>>The split between the "core-config.ini" and multiple backend inis (which was already kind of artificial before this change) now are set in stone by design. There's real reason for doing this.

>Could you clarify this? I'm not sure what you mean. 

I don't like that there are two kinds of ini files, "core-config.ini" and "<frontend_name>-config.ini". I consider this split to be completely artificial, because there's no real reason other than "oh, it has the same contents anyway" and because it incurs at least one of the issues I outlined before (e.g. premising a particular configuration storage location and format - because otherwise, how would you share it between backends?). This is what I meant with "set in stone by design": If we have multiple ini files like this, we give little to no freedom anymore to the frontends as to where configuration files are stored, and how.

>>Not only do I find this design less than ideal, but it also overcomplicates the configuration loading process, which is now split across the core and the GUI.

>Config loading is still controlled by the GUI, but the core config file is still wrapped up in Core. While this may not be ideal, I can't think of a better option that doesn't involve duplicating the options, which IMO is even more overcomplicated.

I disagree that duplicating options is more complicated than having two potentially entirely different configuration subsystems (one in the core, one in the frontend), the latter which also provides room for inconsistency.

>>Mind you, this is done with the expectation that eventually Citra's core should be decoupled from the existing GUIs well enough that other people can start writing frontends for Citra in a clean way. While this is probably utopic from the current standpoint, I don't think we should make explicit misdesigns which will prevent this in the future.

>I totally agree that Citra's core should eventually be decoupled. A few notes though:
>- This doesn't at all prevent Citra's core from being decoupled (the Core already uses the INIH parser, not something specified by the frontend)

As explained above, this forces a particular storage location and storage format for Citra-core-internal configuration upon the frontend developer, though. If I hypothetically were to develop a frontend running on Windows and I'd want to put **all** of the necessary application configuration in the registry, citra's core would instead spam the filesystem with its own stuff.

>- This doesn't at all prevent the startup of new, clean frontends (same reason as above)
"prevent" is a fairly strong word, so I'll agree with you on that one.

>- There's no better place to put a unified configuration loader (Common? Maybe.)
Again, I don't think a unified configuration loader is a good idea, at all.

>---
>I hope we can come to a solution that does not involve that kind of unnecessary duplication, and I would love to hear your suggestions. 

----------------------------------------------------------------------------------------

For what it's worth, I think GLFW provides little enough value that in the long run, we could get rid of it and instead focus on maintaining a single Qt frontend. This is an alternative solution to the duplication issue.

A different alternative, for which we could probably re-use a lot of code in this branch, is to generally let frontends handle all configuration management **in general**, but **optionally** allowing them to use a fallback configuration loader (which would be inih, and which we'd use in GLFW). I would, however, definitely prefer Qt to use its own configuration management (via QSettings), so ultimately it would be no better with regards to duplication.

**Comment by archshift** [comment]
@bunnei, what are your thoughts?

**Comment by bunnei** [comment]
@neobrain brings up some pretty valid points. I propose that we keep the spirit of this PR (remove redundancy),  but instead:
* Get rid of config stuff from Core
* Combine (what was) Core config with the "baseline" frontend config (that GLFW uses), move it to Common and make it just a generic INIH-based config solution that any frontend can use
* Provide a mechanism for frontends to use this OR implement their own backend (e.g. QSettings with Qt), or even just override + extend it
That way, for now the GLFW and Qt frontends could still share the common-config based on INIH, but the design is there to change it when we're ready?

**Comment by archshift** [comment]
Closing for now while I ponder my best course of action from here.

**Pull request closed without merging**